File: 0981-time-based-key-value-store/0981-time-based-key-value-store.py
class TimeMap:

    def __init__(self):
        
        # key - multiple values based on ts
        self.map = defaultdict(list)

    def set(self, key: str, value: str, timestamp: int) -> None:
        self.map[key].append((timestamp, value))

    def get(self, key: str, timestamp: int) -> str:
        if key not in self.map or len(self.map[key])==0:
            return ""
        lst = self.map[key]
        if lst:
            left, right = 0, len(lst)-1
            ans = ""
            while left<=right:
                mid = left+(right-left)//2
                if lst[mid][0]>timestamp:
                    right = mid-1
                else:
                    ans = lst[mid][1]
                    left = mid+1
            return ans
        else:
            return ""


# Your TimeMap object will be instantiated and called as such:
# obj = TimeMap()
# obj.set(key,value,timestamp)
# param_2 = obj.get(key,timestamp)

File: 0981-time-based-key-value-store/0981-time-based-key-value-store.cpp
// SC- O(n) where n is num of times set is called

class TimeMap {
public:

    unordered_map<string, vector<pair<int, string>>> mp; 
    TimeMap() {
    }
    
    void set(string key, string value, int timestamp) {
        // TC- O(1)

        mp[key].push_back({timestamp, value});
    }
    
    string get(string key, int timestamp) {
        // TC- O(logk)

        if (mp.find(key)==mp.end())
            return "";
        
        string ans = "";
        int left =0;
        int right = mp[key].size()-1;
        while(left<=right){
            int mid = left +(right-left)/2;
            if (mp[key][mid].first>timestamp)
                right = mid-1;
            else{
                ans = mp[key][mid].second;
                left = mid+1;
            }
        }
        return ans;
    }
};

/**
 * Your TimeMap object will be instantiated and called as such:
 * TimeMap* obj = new TimeMap();
 * obj->set(key,value,timestamp);
 * string param_2 = obj->get(key,timestamp);
 */